infrastructure/network/addressmanager/addressmanager.go

```diff
@@ -5,1264 +5,254 @@
-	}
-	am.fullNodeNewAddressCount = 0
-	am.fullNodeTriedAddressCount = 0
-}
-
-// HostToNetAddress returns a netaddress given a host address. If
-// the host is not an IP address it will be resolved.
-func (am *AddressManager) HostToNetAddress(host string, port uint16, services appmessage.ServiceFlag) (*appmessage.NetAddress, error) {
-	ip := net.ParseIP(host)
-	if ip == nil {
-		ips, err := am.lookupFunc(host)
-		if err != nil {
-			return nil, err
-		}
-		if len(ips) == 0 {
-			return nil, errors.Errorf("no addresses found for %s", host)
+	if bestAddress == nil {
+		// Send something unroutable if nothing suitable.
+		var ip net.IP
+		if !IsIPv4(remoteAddress) {
+			ip = net.IPv6zero
+		} else {
+			ip = net.IPv4zero
 		}
-		ip = ips[0]
+		services := appmessage.SFNodeNetwork | appmessage.SFNodeBloom
+		bestAddress = appmessage.NewNetAddressIPPort(ip, 0, services)
 	}
 
-	return appmessage.NewNetAddressIPPort(ip, port, services), nil
-}
-
-// NetAddressKey returns a key in the form of ip:port for IPv4 addresses
-// or [ip]:port for IPv6 addresses for use as keys in maps.
-func NetAddressKey(netAddress *appmessage.NetAddress) AddressKey {
-	port := strconv.FormatUint(uint64(netAddress.Port), 10)
-
-	return AddressKey(net.JoinHostPort(netAddress.IP.String(), port))
+	return bestAddress
 }
 
-// GetAddress returns a single address that should be routable. It picks a
-// random one from the possible addresses with preference given to ones that
-// have not been used recently and should not pick 'close' addresses
-// consecutively.
-func (am *AddressManager) GetAddress() *KnownAddress {
-	// Protect concurrent access.
+// Ban marks the given address as banned
+func (am *AddressManager) Ban(address *appmessage.NetAddress) error {
 	am.mutex.Lock()
 	defer am.mutex.Unlock()
 
-	triedAddressBucketArray := am.triedAddressBucketArray(am.localSubnetworkID)
-	triedAddressCount := am.triedAddressCount(am.localSubnetworkID)
-	newAddressBucketArray := am.newAddressBucketArray(am.localSubnetworkID)
-	newAddressCount := am.newAddressCount(am.localSubnetworkID)
-	knownAddress := am.getAddress(triedAddressBucketArray, triedAddressCount, newAddressBucketArray, newAddressCount)
-
-	return knownAddress
-
-}
-
-// getAddress returns a single address that should be routable.
-// See GetAddress for further details.
-func (am *AddressManager) getAddress(triedAddressBucketArray *triedAddressBucketArray, triedAddressCount int,
-	newAddressBucketArray *newAddressBucketArray, newAddressCount int) *KnownAddress {
-
-	// Use a 50% chance for choosing between tried and new addresses.
-	var bucketArray addressBucketArray
-	if triedAddressCount > 0 && (newAddressCount == 0 || am.random.Intn(2) == 0) {
-		bucketArray = triedAddressBucketArray
-	} else if newAddressCount > 0 {
-		bucketArray = newAddressBucketArray
-	} else {
-		// There aren't any addresses in any of the buckets
+	key := netAddressKey(address)
+	if address, ok := am.addresses[key]; ok {
+		address.isBanned = true
```

Comment:
It is idiomatic in go to have the error-case inside the if.
Among other things, the compiler uses this assumption to do branch-prediction.

This means you'll have to declare `var address` outside the if, but it's still preferable.

(Same for Unban, IsBanned)